radareorg/cutter
# Common behaviour for lists with items that have address

**Detailed description**

Add

* AddressableItemContextMenu - basic menu that can be used for most places that have something with address. Currently contains "show in", "copy adderess", "show X-refs", add comment. 
* AdressableItemModel - Qt Item Model with interface for asking item's address and optionally name. I considered using custom role but chose interface with virtual methods instead as it allows enforcing that interface is implemented at compile time. Manually tracking if all the models use correct role would be more fragile.
* ListDockWidget - CutterDockWidget containing tree view and optionally QuckSearch and statusbar with count. Accepts AddressableItemModel and Automatically adds AdressableItemContextMenu. This one saves a lot of duplicate code and provides more consistent behavior. A specific DockWidget implementation needs to provide model which extends AddressableItemModel if necessary customize tree view widget look and add extra context menu actions specific to this widget.

Convert following widgets to use ListDockWidget
* CommentsWidget
* ImportsWidget
* RelocsWidget
* SymbolsWidget
* SegmentsWIdget
* SectionsWIdget
* HeadersWIdget
* FunctionWidget

Some of the provided partial context menu or no context menu. Widget specific context menu entries should be still there.

Bugfixes and existing feature cleanup:
* Escape shortcut for canceling quick filter was broken due to VtablesWidget
* Selecting multiple functions in FunctionWIdget and using "Undefine" action undefines all of them
* Horizontal/Vertical mode selection now uses ActionGroup for managing exclusive selection instead of doing it manually.

**Test plan (required)**

Using context menu
* copy address
* show in
* show x-refs
  * make sure name is shown when applicable and address otherwise
* add comment
* any widget specific context menu items
* action shortcuts (with correct target)
* correct target when opening menu using keyboard

Quick filter
* Ctrl-F focuses on Quick Filter opening it if necesarry
* Quick filter works
* Escape clears the filter

Horizontal/Vertical mode for Functions and Comment widgets
* Qt displays it using Radio button style instead of checkboxes
* Tree item content switches between tree and table
* In tree mode copying adress takes function address in function widget, function or comment address in comment widget

Test at least one action in all widgets 
* CommentsWidget
* ImportsWidget
* RelocsWidget
* SymbolsWidget
* SegmentsWIdget
* SectionsWIdget
  * graphic bars show up
* HeadersWIdget
* FunctionWidget

Make sure it is possible to copy any cell in table view by clicking it and pressing Ctrl-C

** Future work  for part 2**
Deal with rest of of the widgets, probably split ListDockWidget so that TreeView with addressable item model can be used in docks that have more custom gui  like Search. 

Use of AdressableItemContextMenu in places that aren't table row, for example stack widget address and data.

Add comment should probably show existing comment for editing instead of silently replacing it. Current behavior is copied either for Comments or Functions widget.

**Closing issues**

Covers ~~most~~ ~half of #923 . ~~Need to evaluate once finished.~~
